File: java/programmers/1단계/src/모의고사.java
public class 모의고사 {
    public int[] solution(int[] answers) {
        int[] one = {1, 2, 3, 4, 5};
        int[] two = {2, 1, 2, 3, 2, 4, 2, 5};
        int[] three = {3, 3, 1, 1, 2, 2, 4, 4, 5, 5};
        int[] solve = {0, 0, 0};
        
        // 1. 맞춘 문제의 갯수를 solve배열에 저장
        for (int i = 0; i < answers.length; i++) {
            if(one[i % one.length] == answers[i]) solve[0]++;
            if(two[i % two.length] == answers[i]) solve[1]++;
            if(three[i % three.length] == answers[i]) solve[2]++;
        }
        // 2. 최대로 맞춘 문제의 갯수
        int max_value = 0;
        for (int value : solve) {
            if (max_value < value) {
                max_value = value;
            }
        }

        // 3. 최대로 맞춘 인원의 수
        int size = 0;
        for (int k : solve) {
            if (k == max_value) {
                size++;
            }
        }
        // 4. 맞춘 인원을 담을 배열생성
        int[] answer = new int[size];
        int num = 0;
        for (int j = 0; j < solve.length; j++) {
            if(max_value == solve[j]){
                answer[num++] = j+1;
            }
        }
        return answer;
    }

    public static void main(String[] args) {
        모의고사 sol = new 모의고사();
        int[] answers = {1,2,3,4,5};
        int[] solution = sol.solution(answers);

        for (int i : solution) {
            System.out.println("i = " + i);
        }
    }
}

File: python/python_Programmers/1단계/모의고사.py
# 내풀이
def solution(answers):
    answer = []

    list1 = [1, 2, 3, 4, 5]
    list2 = [2, 1, 2, 3, 2, 4, 2, 5]
    list3 = [3, 3, 1, 1, 2, 2, 4, 4, 5, 5]

    solve = [0, 0, 0]
    for i in range(0, len(answers)):
        if list1[i % len(list1)] == answers[i]:
            solve[0] += 1
        if list2[i % len(list2)] == answers[i]:
            solve[1] += 1
        if list3[i % len(list3)] == answers[i]:
            solve[2] += 1
    max_value = max(solve)

    for j in range(0, len(solve)):
        if solve[j] == max_value:
            answer.append(j+1)
    return answer

print(solution([1,3,2,4,2]))

# 모범 풀이
from itertools import cycle

def solution2(answers):
    giveups = [
        cycle([1,2,3,4,5]),
        cycle([2,1,2,3,2,4,2,5]),
        cycle([3,3,1,1,2,2,4,4,5,5]),
    ]
    scores = [0, 0, 0]
    for num in answers:
        for i in range(3):
            if next(giveups[i]) == num:
                scores[i] += 1
    highest = max(scores)

    return [i + 1 for i, v in enumerate(scores) if v == highest]
print(solution2([1,3,2,4,2]))

# cycle 활용법
A = [1,2,3]
B = ['A','B','C','D']
C = [10,20,30,40,50]
R_Z = zip(A, B, C)
print(next(R_Z))
print(next(R_Z))
print(next(R_Z))

R_L = zip(cycle(A),cycle(B),cycle(C))
for i in range(0, len(C)):
    print('r_l : ', next(R_L))